File: cercaend/fix_context.py
import re
import os

def fix_context_warnings():
    output_files = {}
    with open('analyze_output.txt', 'r', encoding='utf-8') as f:
        lines = f.readlines()
        
    for line in lines:
        if 'use_build_context_synchronously' in line:
            # Parse line:  info - lib\mainpages\orderpage\orderpage_widget.dart:608:18 - ...
            parts = line.split(' - ')
            if len(parts) >= 3:
                file_info = parts[1].strip()
                if ':' in file_info:
                    filepath, line_num, col_num = file_info.rsplit(':', 2)
                    line_num = int(line_num) - 1 # 0-indexed
                    if filepath not in output_files:
                        output_files[filepath] = set()
                    output_files[filepath].add(line_num)
    
    for filepath, line_nums in output_files.items():
        if not os.path.exists(filepath):
            continue
            
        with open(filepath, 'r', encoding='utf-8') as f:
            lines = f.readlines()
            
        # Process from bottom to top to not mess up line numbers
        for ln in sorted(list(line_nums), reverse=True):
            if ln < len(lines):
                target_line = lines[ln]
                indent = len(target_line) - len(target_line.lstrip())
                indent_str = ' ' * indent
                
                # Check if we already have a mounted check
                already_has_mounted = False
                for i in range(1, 4):
                    if ln - i >= 0:
                        prev_line = lines[ln - i].strip()
                        if 'mounted' in prev_line:
                            already_has_mounted = True
                            break
                        if prev_line and prev_line != '}' and prev_line != '{':
                            break # stop checking if we hit a real line without mounted
                
                if not already_has_mounted:
                    # Insert context.mounted
                    lines.insert(ln, indent_str + 'if (!context.mounted) return;\n')
        
        with open(filepath, 'w', encoding='utf-8') as f:
            f.writelines(lines)
            
    print(f"Fixed files: {len(output_files)}")

if __name__ == '__main__':
    fix_context_warnings()

File: cercaend/fix_context.dart
import 'dart:io';

void main() {
  final file = File('analyze_output_utf8.txt');
  if (!file.existsSync()) {
    print('analyze_output_utf8.txt not found');
    return;
  }

  final lines = file.readAsLinesSync();
  final outputFiles = <String, Set<int>>{};

  for (var line in lines) {
    if (line.contains('use_build_context_synchronously')) {
      final parts = line.split(' - ');
      if (parts.length >= 3) {
        final fileInfo = parts[1].trim();
        final fileParts = fileInfo.split(':');
        if (fileParts.length >= 3) {
          final filepath = fileParts[0];
          final lineNum = int.tryParse(fileParts[1]) ?? 0;
          if (lineNum > 0) {
            outputFiles.putIfAbsent(filepath, () => <int>{});
            outputFiles[filepath]!.add(lineNum - 1);
          }
        }
      }
    }
  }

  for (var entry in outputFiles.entries) {
    var filepath = entry.key;
    final dartFile = File(filepath);
    if (!dartFile.existsSync()) {
      print('File not found: $filepath');
      continue;
    }

    final dartLines = dartFile.readAsLinesSync();
    final lineNumsDesc = entry.value.toList()..sort((a, b) => b.compareTo(a));

    for (var ln in lineNumsDesc) {
      if (ln < dartLines.length) {
        final targetLine = dartLines[ln];
        final indentMatch = RegExp(r'^\s*').firstMatch(targetLine);
        final indentStr = indentMatch?.group(0) ?? '';

        bool alreadyHasMounted = false;
        for (var i = 1; i <= 3; i++) {
          if (ln - i >= 0) {
            final prevLine = dartLines[ln - i].trim();
            if (prevLine.contains('mounted')) {
              alreadyHasMounted = true;
              break;
            }
            if (prevLine.isNotEmpty && prevLine != '}' && prevLine != '{') {
              break;
            }
          }
        }

        if (!alreadyHasMounted) {
          var insertStr = '${indentStr}if (!context.mounted) return;';
          dartLines.insert(ln, insertStr);
        }
      }
    }

    dartFile.writeAsStringSync('${dartLines.join('\n')}\n');
    print('Fixed: $filepath');
  }
}
